adapters/zio-http/src/main/scala/caliban/ZHTTPAdapter.scala

```diff
@@ -0,0 +1,240 @@
+    enableIntrospection: Boolean,
+    queryExecution: QueryExecution
+  ): URIO[R, Json] =
+    interpreter
+      .executeRequest(
+        request,
+        skipValidation = skipValidation,
+        enableIntrospection = enableIntrospection,
+        queryExecution
+      )
+      .foldCause(cause => GraphQLResponse(NullValue, cause.defects).asJson, _.asJson)
+
+  implicit class HttpErrorOps[R, E <: Throwable, A](private val zio: ZIO[R, io.circe.Error, A]) extends AnyVal {
+    def handleHTTPError: ZIO[R, HttpError, A] = zio.mapError({
+      case DecodingFailure(error, _)  =>
+        HttpError.BadRequest.apply(s"Invalid json: $error")
+      case ParsingFailure(message, _) =>
+        HttpError.BadRequest.apply(message)
+      case t: Throwable               => HttpError.InternalServerError.apply("Internal Server Error", Some(t.getCause()))
+    })
+  }
+
+  private val protocol = SocketProtocol.subProtocol("graphql-ws")
+
+  private val keepAlive = (keepAlive: Option[Duration]) =>
+    keepAlive match {
+      case None           => ZStream.empty
+      case Some(duration) =>
+        ZStream
+          .succeed(Text("""{"type":"ka"}"""))
+          .repeat(Schedule.spaced(duration))
+    }
+
+  private val connectionError                            = ZStream.succeed(Text("""{"type":"connection_error"}"""))
+  private val connectionAck                              = ZStream.succeed(Text("""{"type":"connection_ack"}"""))
+  private val close                                      = ZStream.succeed(WebSocketFrame.close(1000))
+  private def toStreamError[E](id: Option[String], e: E) = ZStream.succeed(
+    Text(
+      Json
+        .obj(
+          "id"      -> Json.fromString(id.getOrElse("")),
+          "type"    -> Json.fromString("error"),
+          "message" -> Json.fromString(e.toString())
+        )
+        .toString()
+    )
+  )
+
+  private def toResponse[E](id: String, fieldName: String, r: ResponseValue, errors: List[E]): WebSocketFrame.Text =
+    toResponse(
+      id,
+      GraphQLResponse(
+        ObjectValue(List(fieldName -> r)),
+        errors
+      )
+    )
+
+  private def toResponse[E](id: String, r: GraphQLResponse[E]) = Text(
+    Json
+      .obj("id" -> Json.fromString(id), "type" -> Json.fromString("data"), "payload" -> r.asJson)
+      .toString()
+  )
+
+  type Subscriptions = Ref[Map[String, Promise[Any, Unit]]]
+
+  private def trackSubscription(id: String, subs: Subscriptions) =
+    ZStream
+      .fromEffect(for {
+        p <- Promise.make[Any, Unit]
+        _ <- subs.update(m => m.updated(id, p))
+      } yield p)
+
+  private def removeSubscription(id: Option[String], subs: Subscriptions) =
+    ZStream
+      .fromEffect(
+        id.map(id =>
+          subs
+            .modify(map => (map.get(id), map - id))
+            .flatMap { p =>
+              IO.whenCase(p) { case Some(p) =>
+                p.interrupt
+              }
+            }
+        ).getOrElse(ZIO.unit)
```

Comment:
you can use `ZIO.whenCase(id) { case Some(id) => }` which is a bit shorter and avoids `getOrElse(ZIO.unit)`